File: backend-gin/internal/lib/clients/huggingFace.go
package clients

import (
	"backend/internal/lib/utils"
	"encoding/json"
	"fmt"
	"net/http"
)

type HuggingFaceClientConfig struct {
	Token      string
	EmbedModel string
	Timeout    int
}

type HuggingFaceClient struct {
	config HuggingFaceClientConfig
}

func (h *HuggingFaceClient) BaseUrl() string {
	return fmt.Sprintf("https://router.huggingface.co/hf-inference/models/%s/pipeline/feature-extraction", h.config.EmbedModel)
}

func (h *HuggingFaceClient) EmbedText(text string) ([]float64, error) {
	requestBody := map[string]any{"inputs": []string{text}}
	url := h.BaseUrl()
	resp, err := utils.PostRequest(
		url, h.config.Timeout, requestBody, h.config.Token,
	)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API request failed with status: %s", resp.Status)
	}

	var embeddingResponse [][]float64
	if err := json.NewDecoder(resp.Body).Decode(&embeddingResponse); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if len(embeddingResponse) == 0 || len(embeddingResponse[0]) == 0 {
		return nil, fmt.Errorf("empty embedding response")
	}
	return embeddingResponse[0], nil
}

func NewHuggingFaceClient(config HuggingFaceClientConfig) HuggingFaceClient {
	if config.EmbedModel == "" {
		config.EmbedModel = "sentence-transformers/all-MiniLM-L6-v2"
	}
	return HuggingFaceClient{config: config}
}

File: backend-axum/src/lib_/clients/hugging_face.rs
use reqwest::{Client, Error as ReqwestError, Response};
use std::time::Duration;

use crate::lib_::types_::{ApiError, SimpleError};

#[derive(Debug, Clone)]
pub struct HuggingFaceClientConfig {
    pub token: String,
    pub embed_model: String,
    pub timeout: usize,
}

pub struct HuggingFaceClient {
    client: Client,
    config: HuggingFaceClientConfig,
}

impl HuggingFaceClient {
    pub async fn new(config: HuggingFaceClientConfig) -> Result<Self, ReqwestError> {
        let client = Client::builder()
            .timeout(Duration::from_secs(config.timeout as u64))
            .build()?;

        let embed_model = if config.embed_model.is_empty() {
            "sentence-transformers/all-MiniLM-L6-v2".to_string()
        } else {
            config.embed_model.clone()
        };

        let config = HuggingFaceClientConfig {
            embed_model,
            ..config
        };

        Ok(Self { client, config })
    }

    fn base_url(&self) -> String {
        format!(
            "https://router.huggingface.co/hf-inference/models/{}/pipeline/feature-extraction",
            self.config.embed_model
        )
    }

    pub async fn embed_text(&self, text: &str) -> Result<Vec<f32>, ApiError> {
        if text.is_empty() {
            return Err(ApiError::internal_error(
                "text cannot be empty",
                Box::new(SimpleError::from("text cannot be empty")),
            ));
        }

        let url = self.base_url();
        let body = serde_json::json!({ "inputs": [text] });

        let resp: Response = self
            .client
            .post(&url)
            .bearer_auth(&self.config.token)
            .json(&body)
            .send()
            .await
            .map_err(|e| {
                ApiError::failed_dependency("embedding server could not be reached", Box::new(e))
            })?;

        if !resp.status().is_success() {
            return Err(ApiError::failed_dependency(
                "embedding server failed",
                Box::new(SimpleError::from(format!(
                    "server returned {} HTTP response",
                    resp.status()
                ))),
            ));
        }

        let embedding_response: Vec<Vec<f32>> = resp.json().await.map_err(|e| {
            ApiError::failed_dependency("failed to parse embedding server response", Box::new(e))
        })?;

        if embedding_response.is_empty() || embedding_response[0].is_empty() {
            return Err(ApiError::failed_dependency(
                "server response did not return embedding vector",
                Box::new(SimpleError::from("server response empty")),
            ));
        }

        Ok(embedding_response[0].clone())
    }

    pub async fn close(self) -> Result<(), String> {
        // reqwest client drops automatically
        Ok(())
    }
}
